### DISCUSSION TITLE
"photoprism users rm" fails

### DISCUSSION THREAD

**Discussion by marieper** [photoprism/photoprism#4427] [Q&A] [answered]
As I missed the fact that a user is created with admin role as default, I want to recreate that user.

photoprism running in docker.
The default super user works as intended.

```
mapenas:~# docker exec photoprism-photoprism-1 photoprism users rm erinyerna -y
time="2024-07-27T22:07:59Z" level=info msg="Become a member today, support our mission and enjoy our member benefits! 💎"
time="2024-07-27T22:07:59Z" level=info msg="Visit https://www.photoprism.app/membership to learn more."

time="2024-07-27T22:07:59Z" level=info msg="user 'erinyerna' was not deleted"
```

At one point/occasion the command tried to execute, but expected a user input to confirm the deletion of the user, which of course was not possible (and not expected).

What needs to be done in order to delete the user to recreate it?
"photoprism auth ls" lists only one session, that of the super user.

**Comment by graciousgrey** [comment, marked as answer]
The users rm command has no -y flag.

If you want to delete a user without confirmation, you can use -f. 

All our commands also support --help if you are unsure about the usage. 

More information can be found here:
- https://docs.photoprism.app/user-guide/users/cli/

**Comment by marieper** [comment]
Thank you, using the -f flag resolved the issue of denial.

It is unfortunate that the web-documentation does not address the force flag, however, the CLI help text on the "users rm" command does.

That the actual intention of creating an unprivileged user was blocked by the subscription plan in altogether another topic.

**Reply by lastzero** [reply]
You can create regular admins and guest users, which are both less privileged than the default super admin. On top you get free support directly from the maintainers, but that's another topic.

**Comment by sseide** [comment]
Current implementation is very confusing. I ran into the same problem - trying to delete a user and its not working.

Docker based setup:
```
$ docker exec photoprism-photoprism-1 photoprism users ls
time="2024-08-01T12:57:14Z" level=info msg="found 2 users"

|------------------|---------------|-------|----------------|-------------|-----------|---------------------|----------|-------------|
|       UID        |   Username    | Role  | Authentication | Super Admin | Web Login |     Last Login      |  WebDAV  | Upload Path |
|------------------|---------------|-------|----------------|-------------|-----------|---------------------|----------|-------------|
| usg...           | admin         | Admin | Local          | Yes         | Enabled   | 2024-08-01 11:32:57 | Enabled  |             |
| ush...           | stefantest    | Guest | OIDC           | No          | Enabled   | 2024-08-01 11:32:04 | Disabled |             |
|------------------|---------------|-------|----------------|-------------|-----------|---------------------|----------|-------------|

$ docker exec photoprism-photoprism-1 photoprism users rm stefantest

time="2024-08-01T12:57:41Z" level=info msg="user 'stefantest' was not deleted"
```

Why? No hint whats going wrong. Documentation does not even mention the force flag.
https://docs.photoprism.app/user-guide/users/cli/#managing-user-accounts -> "users rm" command has no "[options]", only the "username" as parameter.

The log message should clearly state that it is not deleted and you must run command again with force flag to really delete it.
Similar behavior have some dangerous cli based tools (dry-run as default, e.g file system repair etc.) but they say that it must be forced to be done.

And switching logging to DEBUG/TRACE level or activating DEBUG mode does not help either. No usefull information given why the account is not delete. Please improve application logging to better support people on hunting problems.

**Reply by lastzero** [reply]
See our documentation for how to run commands with `docker` if you - for whatever reason - prefer not to use `docker compose`: https://docs.photoprism.app/getting-started/docker/#examples

Based on this, you should try the following or use the command flag signaling non-interactive approval to confirm the action:
```
docker exec -ti photoprism-photoprism-1 photoprism users rm stefantest
```

`-ti` stands for "terminal interactive’" so you can interactively confirm the action by pressing the "Y" key. This is a standard method to prevent accidental deletion, e.g. due to a typo. This `docker` command flag is also very common and well documented. How can it be so confusing?

To display the command help, you can use the `--help` flag (as documented), e.g. `photoprism users rm --help`:

```
NAME:
   photoprism users rm - Deletes a registered user account

USAGE:
   photoprism users rm [command options] [username]

OPTIONS:
   --force, -f  don't ask for confirmation
```

**Note the `--force` flag, which prevents a confirmation prompt, e.g. if you want to run the command with Docker in non-interactive mode.**

**Reply by lastzero** [reply]
Here is the corresponding section in the official Docker documentation:
https://docs.docker.com/reference/cli/docker/container/exec/

**Reply by sseide** [reply]
https://docs.photoprism.app/user-guide/users/cli/#managing-user-accounts
![Screenshot_20240801_160248](https://github.com/user-attachments/assets/71c091e1-b6cc-4d8b-ad30-72d3a0d33642)
This part of the documentation listing all commands to manage user does not show that there might options exists for "rm".

Therefor no use of running the "rm --help" command to query possible options.
This documentation should be updated then.

**Reply by sseide** [reply]
I know about the interactive flag - but running your commands without it does not give any hint at all that interactive feedback might be needed. Or that it is running in a special non-interactive mode due to missing TTY. Here the output of all commands needing confirmation should be changed to at least state so.

And having either new admins or admin not interacting with the system for half a year might not know/remember that "-it" is crucial here. And as said - even tracing does not give a hint whats going wrong.

Here command and log output can be greatly be improved.

**Reply by lastzero** [reply]
It's not that unusual to require confirmation when deleting something, and the command help shows the existence of a `--force` flag that prevents asking for confirmation - so you'd think confirmation would be required otherwise?

The [documentation](https://docs.photoprism.app/getting-started/docker/) also clearly recommends using Docker Compose, which does not require this flag (which you should otherwise use by default if, for whatever reason, you prefer to use the `docker` command):

> We recommend using [Docker Compose](https://docs.photoprism.app/getting-started/docker-compose/) because it is easier and provides more convenience for running multiple services than the [pure Docker command-line interface](https://docs.docker.com/engine/reference/commandline/cli/).

If anything, you should write to the Docker team that you want a warning displayed if you forgot to use the `-ti` flag, as this is not our software and we do not provide support for it as part of our free community support. Thank you for your consideration!

**Reply by lastzero** [reply]
> This documentation should be updated then.

This is the pure command documentation in case you are using a regular terminal. The same documentation says that you must otherwise prefix the command with `docker compose exec [service] ...` (or, if you are an experienced user who prefers to use pure Docker, `docker exec -ti`, which we do not repeat in all sections of our end-user documentation so as not to confuse less experienced users who follow our recommendations):

> Note that when you use [Docker Compose](https://docs.photoprism.app/getting-started/docker-compose/#command-line-interface) and do not [already have a terminal session open](https://docs.photoprism.app/getting-started/docker-compose/#opening-a-terminal), you must prepend docker compose exec photoprism so that the command is executed within the photoprism container, for example:
> 
> ```bash
> docker compose exec photoprism photoprism passwd admin
> ```
> 
> This also [applies to other terminal commands](https://docs.photoprism.app/getting-started/docker-compose/#examples), including those listed below.

**Reply by sseide** [reply]
Thanks - learned something new today. I did not knew that docker compose uses interactive sessions as default. Using it to seldom as main work for years is done with kubectl and docker command...

**Reply by sseide** [reply]
Regarding the stdout or log messages running commands non-interactive.

Part of the problem it seems is "manifoldco/promptui" library not checking if its running inside a TTY. The information is available from "chzyer/readline" used internally. 

And another part is your code catching the error but silently discarding it. To help users of the software errors should either be logged or returned/thrown for others to handle but not silently dropped.
https://github.com/photoprism/photoprism/blob/6c1dd02c2216617b1a19eb04ca820ff6740cea70/internal/commands/users_remove.go#L62-L65
From an admin point of view this could make the difference between looking 5min into command output and log files or searching a day for the real root cause of a problem. To best help your users please log all errors and not just throw them away.

**Reply by lastzero** [reply]
If we had a huge team of developers, I would have it improved for you right away. However, our core team is on vacation until August 19 and we already have a lot of other issues to take care of then - so there's no way to get this done and properly tested in the short term. Keep in mind that breaking it would be even worse than not improving it for your use case. And at least we have documentation showing how to use it as intended, e.g. with the `-it` flag when running the command with `docker exec`.

**Reply by sseide** [reply]
No need to hurry here, nothing critical. And due to the existing issue report and your very fast responses the problems/questions where solved very fast.

And just see it this way - there is a way to make your good software better. And every bit of helpful information you give your users with logs and program output helps them to solve their problems themself. Due to it they do not need to open an issue here, freeing time on your side to develop new functions. In the end a win-win for all.

Have a good holiday!
